### PR TITLE
[Core][deprecate run_function_on_all_workers 3/n] delete run_function_on_all_workers

### PR DESCRIPTION
**Pull request by scv119**
## Why are these changes needed?

This function is deprecated. The only use of run_funcion_on_all_workers in ray core has been replaced by https://github.com/ray-project/ray/pull/31383
 Delete it to make our worker prestart change simpler.
This PR depends on #31383 #31528

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/_private/worker.py
Comment (rkooo567) [inline python/ray/_private/worker.py:691]: why don't we raise a deprecation exception for now? Not sure about the deprecation policy, but it could feel like a sudden deprecation because we don't print warning on `@Deprecated`

File: python/ray/_private/import_thread.py
Comment (rkooo567) [inline python/ray/_private/import_thread.py:160]: Does it mean we are losing this data? This is kind of important for the timeline feature.

**Comment by pcmoritz** [comment]
Very excited for this change, thank you so much for making it! This is code that is still around from a long time ago :)

**Comment by scv119** [comment]
Will continue this after Ray 2.3.

File: python/ray/_private/worker.py
Comment (clarng) [inline python/ray/_private/worker.py:2962]: revert?

File: python/ray/_private/worker.py
Comment (clarng) [reply python/ray/_private/worker.py:696]: +1 to print exception (and maybe point to runtime env doc)

**Review by clarng — APPROVED** [review]

**Comment by edoakes** [comment]
🤩

File: python/ray/_private/worker.py
Comment (scv119) [reply python/ray/_private/worker.py:696]: yeah, we have already done it in 2.3 :) this PR is for 2.4

**Comment by scv119** [comment]
rebase

**Review by rkooo567 — COMMENTED** [review]
Hmm is it really a good idea to deprecate this without the proper alternative? I feel like we need sth like a callback API that's called when the worker starts?

**Comment by scv119** [comment]
> I feel like we need sth like a callback API that's called when the worker starts?

I thought that API called runtime env :)

**Comment by rkooo567** [comment]
Hmm I actually have 2 suggestions in this case?

https://docs.ray.io/en/master/ray-core/handling-dependencies.html -> I saw this doc, and it is not clear to me where to look in order to perform the same action. Can you link to the sub-category that has the exact API to replace this? 

Instead of completely removing the function, why don't we keep the function but just raise an exception (for the backward compatibility)?

**Comment by rkooo567** [comment]
Unstale. We can replace this to https://docs.google.com/document/d/1ngiuAZAMnl9c4LozoTpWh37KPviDRIpmEjEI6BsNL7w/edit# by ray 2.5

File: python/ray/tests/test_cancel.py
Comment (pcmoritz) [inline python/ray/tests/test_cancel.py:281]: This looks like a leftover from some previous commenting out :)

**Review by pcmoritz — APPROVED** [review]

**Comment by pcmoritz** [comment]
There is a modin related test failure.

Looks like we got lucky here and modin's dependency on run_functions_on_all_workers was removed here: https://github.com/modin-project/modin/pull/4603 so maybe bumping the modin version will fix this? I hope that's true, otherwise there will be pain.

**Comment by rkooo567** [comment]
cc @RehanSD for the awareness.

**Review by rkooo567 — APPROVED** [review]
Actually @pcmoritz instead of completely removing the API, why don't we raise an exception? Or do you think it is better just removing it? If so, we should make sure this is documented in 2.6 (do we have any place to track deprecation somewhere?)

**Comment by rkooo567** [comment]
Also, I will make sure the replacement API will be available and documented by ray 2.6

**Comment by pcmoritz** [comment]
Given this was a private API and has been deprecated for a while, I think it is fine to just remove it. But it is a good idea to mention in the release notes the worker process init hook you are working on :)

**Comment by pcmoritz** [comment]
Looks like one more lint is failing. I'll also try to get the remaining code owner approvals :)

**Comment by pcmoritz** [comment]
There is another failure here: https://buildkite.com/ray-project/oss-ci-build-pr/builds/24959#01889a90-c54a-46c9-af89-0b62df111d16 :)

**Review by ericl — APPROVED** [review]
Stamping

**Comment by scv119** [comment]
huh what a coincidence https://github.com/ray-project/ray/commit/369f68eade807133710648cb021903acdfe73f06#diff-41539696967c12463a48d082cec221167ea1b96ec702641db33479116c02297b

**Comment by pcmoritz** [comment]
Haha nice :)

**Pull request merged**
